python/cuspatial/cuspatial/core/binops/binops.py

```diff
@@ -0,0 +1,409 @@
+            The left-hand-side of the binary operation.
+        rhs : GeoSeries
+            The right-hand-side of the binary operation.
+        align : bool
+            If True, align the indices of `lhs` and `rhs` before performing
+            the binary operation. If False, `lhs` and `rhs` must have the
+            same index.
+
+        Returns
+        -------
+        GeoSeries
+            A GeoSeries containing the result of the binary operation.
+        """
+        (self.lhs, self.rhs) = lhs.align(rhs) if align else (lhs, rhs)
+        self.align = align
+
+        # Type disambiguation
+        # Type disambiguation has a large effect on the decisions of the
+        # algorithm.
+        (lhs, rhs, indices) = self.preprocess(op, self.lhs, self.rhs)
+
+        # Binop call
+        _binop = getattr(self, op)
+        point_result = _binop(lhs, rhs)
+
+        # Postprocess: Apply discrete math rules to identify relationships.
+        final_result = self.postprocess(op, indices, point_result)
+
+        # Return
+        self.op_result = final_result
+
+    def __call__(self) -> cudf.Series:
+        """Return the result of the binary operation."""
+        return self.op_result
+
+    def contains_properly(self, lhs, points):
+        """Compute the contains_properly relationship between two GeoSeries.
+        A feature A contains another feature B if no points of B lie in the
+        exterior of A, and at least one point of the interior of B lies in the
+        interior of A. This is the inverse of `within`."""
+        if contains_only_points(lhs) and contains_only_points(points):
+            return self.equals(lhs, points)
+        if not contains_only_polygons(lhs):
+            raise TypeError(
+                "`.contains` can only be called with polygon series."
+            )
+
+        # call pip on the three subtypes on the right:
+        point_result = contains_properly(
+            points.x,
+            points.y,
+            lhs.polygons.part_offset[:-1],
+            lhs.polygons.ring_offset[:-1],
+            lhs.polygons.x,
+            lhs.polygons.y,
+        )
+        return point_result
+
+    def equals(self, lhs, rhs):
+        """Compute the equals relationship between two GeoSeries."""
+        if contains_only_points(lhs):
+            result = lhs.points.xy.equals(rhs.points.xy)
+        elif contains_only_linestrings(lhs):
+            result = lhs.lines.xy.equals(rhs.lines.xy)
+        elif contains_only_polygons(lhs):
+            result = lhs.polygons.xy.equals(rhs.polygons.xy)
+        elif contains_only_multipoints(lhs):
+            result = lhs.multipoints.xy.equals(rhs.multipoints.xy)
+        return result
+
+    def touches(self, lhs, rhs):
+        """Compute the touches relationship between two GeoSeries:
+        Two objects touch if they share at least one point in common, but their
+        interiors do not intersect."""
+        return self.equals(lhs, rhs)
```

Comment:
Eh, I don't see preprocess and postprocess to touches. I think it takes more than just `equals`...

Reply:
This is essentially crust - at one time I thought I would write some cases for `touches` but found that equality doesn't work for any cases of `touches`. I left it in because I'll be writing it soon.

Reply:
Also this will appear correctly supported in the next PR https://github.com/rapidsai/cuspatial/issues/838

Reply:
Removed